File: src/vm/virtualmachine.py
'''Implements VM and instructions'''

from __future__ import print_function

try:
    input = raw_input
except:
    pass

import opset
import operator
from debuggable import Debuggable

class VirtualMachine(Debuggable):
    def __init__(self, debug=False):
        '''Build stack, memory, and instruction set'''
        self.stack = []
        self.memory = []
        self.pc = 0
        self.build()
        Debuggable.__init__(self, "[VM]", debug)

    def build(self):
        '''Build instruction set'''

        # Build nullary (stack) operations
        def dup():
            '''Duplicate top of stack'''
            self.stack.append(self.stack[-1])

        def rot():
            '''Rotate top two stack elements'''
            self.stack[-1], self.stack[-2] = self.stack[-2], self.stack[-1]

        def stackop2(function):
            """Turns a binary function into a binary stack operation"""
            import functools
            @functools.wraps(function)
            def op():
                x = self.stack.pop()
                self.stack[-1] = int(function(self.stack[-1], x))
            return op

        operator.div = operator.floordiv

        # Nullary jump table
        self.nullary = [
            stackop2(operator.add),
            stackop2(operator.sub),
            stackop2(operator.mul),
            stackop2(operator.div),
            stackop2(operator.lt),
            stackop2(operator.gt),
            stackop2(operator.eq),
            stackop2(operator.le),
            stackop2(operator.ge),
            stackop2(operator.and_),
            stackop2(operator.or_),
            stackop2(operator.xor),
            dup,
            rot,
        ]

        # Build unary (memory) operations
        def lit(literal):
            """Push literal on stack"""
            self.stack.append(literal)

        def load(address):
            """Put data on top of stack"""
            self.stack.append(self.memory[address])

        def save(address):
            """Save data from top of stack to memory"""
            self.memory[address] = self.stack.pop()

        def out(address):
            """Print data from memory"""
            print(self.memory[address])

        def read(address):
            """Read data into memory"""
            self.memory[address] = int(input("[]> "))

        def jmp(address):
            '''Unconditional jump'''
            self.pc = address

        def jnz(address):
            '''Jump if top of stack is nonzero'''
            if self.stack[-1]:
                self.pc = address

        # Jump table for unary instructions
        self.unary = [
            lit,
            load,
            save,
            out,
            read,
            jmp,
            jnz,
        ]

    def run(self, program):
        '''Execute list of instructions'''
        end = program[0] # First opcode is length
        self.memory = [0] * program[1] # Second opcode is memory size requirement
        self.pc = 2

        while self.pc < end and program[self.pc] != opset.HALT:
            instruction = program[self.pc]
            self.pc += 1
            if opset.nullary(instruction):
                self.debug("%02d %s", self.pc, opset.byte2name[instruction])
                self.nullary[instruction]()
            elif opset.unary(instruction):
                argument = program[self.pc]
                self.pc += 1
                self.debug("%02d %s %s", self.pc, opset.byte2name[instruction], argument)
                self.unary[instruction - opset.LIT](argument)

if __name__ == "__main__":
    from sys import argv
    from assembler import Assembler
    from argparser import parseArgs

    # Grab command line arguments
    args = parseArgs()

    try:
        asm = Assembler(args.debug)
        vm = VirtualMachine(args.debug)
        vm.run(asm.assemble(open(args.filename)))
    except Exception as e:
        print(e)

# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4

File: src/vm/virtualmachine.cpp
#include <iostream>
#include <vector>
#include <stdarg.h>

#include "opset.h"

enum Result { SUCCESS, FAILURE };

// Get next opcode, increment pc
#define NEXTOP (opcode)program[pc]; pc += 1

// Get argument, increment pc
#define NEXTARG program[pc]; pc += 1

// Turn binary operator into stack op
#define STACKOP(op) \
    x = stack.back(); \
    stack.pop_back(); \
    stack[stack.size() - 1] = (int)(stack.back() op x)

// Push literal value on stack
#define IMMED(val) stack.push_back(val)

// Dereference memory and push on stack
#define PUSH(addr) IMMED(memory[addr])

// Pop value from stack and save to memory address
#define POP(addr) memory[addr] = stack.back(); stack.pop_back();

// Halt and report failure on unrecognized opcode
#define BADOP \
    std::cout << "Unrecognized instruction " << instruction << " at " << pc << std::endl; \
    result = FAILURE; \
    goto halt

// Print stack for debugging
#define STACKPRINT \
    std::cout << "["; \
    for(int i = 0; i < stack.size(); ++i) { \
        std::cout << stack[i] << " ";\
    }\
    std::cout << "]" << std::endl

// Debug printf
#ifdef DEBUG
void debug(const char *format, ...){
    va_list args;
    va_start(args, format);
    vprintf(format, args);
    va_end(args);
}
#else
#define debug(...)
#endif

// Execute opcodes
int run(int *program) {

    // First opcode is length
    int length = program[0];

    // Second opcode is memory requirement
    std::vector<int> memory(program[1], 0);

    // Used for arithmetic and logical operations and jumps
    std::vector<int> stack;

    // program counter
    int pc = 2;

    // Used for calculations
    int x = 0;

    // Success means we finished the program without hitting a bad opcode
    Result result = SUCCESS;

    while(pc < length) {
        opcode instruction = NEXTOP;
        if(nullary(instruction)) {
            debug("%02d %s\n", pc, byte2name[instruction]);
            switch(instruction) {
                case ADD: STACKOP(+);  break;
                case SUB: STACKOP(-);  break;
                case MUL: STACKOP(*);  break;
                case DIV: STACKOP(/);  break;
                case LT:  STACKOP(<);  break;
                case GT:  STACKOP(>);  break;
                case EQ:  STACKOP(==); break;
                case LE:  STACKOP(<=); break;
                case GE:  STACKOP(>=); break;
                case AND: STACKOP(&);  break;
                case OR:  STACKOP(|);  break;
                case XOR: STACKOP(^);  break;
                case DUP:
                    stack.push_back(stack.back());
                    break;
                case ROT:
                    x = stack[stack.size() - 1];
                    stack[stack.size() - 1] = stack[stack.size() - 2];
                    stack[stack.size() - 2] = x;
                    break;
                default:
                    BADOP;
            }
        } else if(unary(instruction)) {
            int arg = NEXTARG;
            debug("%02d %s %d\n", pc, byte2name[instruction], arg);
            switch(instruction) {
                case LIT: IMMED(arg); break;
                case LOAD: PUSH(arg); break;
                case SAVE: POP(arg); break;
                case OUT:
                    std::cout << memory[arg] << std::endl;
                    break;
                case IN:
                    std::cout << "[]> ";
                    std::cin >> memory[arg];
                    break;
                case JNZ:
                    if(!stack.back())
                        break;
                case JMP:
                    pc = arg;
                    break;
                case NOP:
                    break;
                case HALT:
                    goto halt;
                default:
                    BADOP;
            }
        } else {
            BADOP;
        }
    }
halt:
    return result;
}

int main(int argc, char *argv[]) {
    if(argc < 2) {
        std::cout << "Must provide filename" << std::endl;
        return 2;
    }

    FILE *file = fopen(argv[1], "rb");
    if(file == NULL) {
        std::cout << "Could not read file '" << argv[1] << "'" << std::endl;
        return -1;
    }

    int length = 0;
    if(fread(&length, sizeof(int), 1, file) != 1 || length <= 0) {
        std::cout << "Malformed binary file" << std::endl;
        return -1;
    }

    int *program = new int[length];
    program[0] = length;
    if(fread(program + 1, sizeof(int), length - 1, file) < length - 1) {
        std::cout << "Incorrect length specification" << std::endl;
        return -1;
    }

    run(program);
    delete[] program;

    return 0;
}

